web/client-api/src/main/java/io/deephaven/web/client/api/tree/JsTreeTable.java

```diff
@@ -209,455 +232,454 @@ public TreeRow(int offsetInSnapshot, Object[] dataColumns, Object rowStyleColumn
+                        TreeViewportData vd = new TreeViewportData(
+                                offset,
+                                includedRows.isEmpty() ? 0 : includedRows.size(),
+                                snapshot.getTableSize(),
+                                snapshot.getDataColumns(),
+                                queryColumns);
+
+                        handleUpdate(nextSort, nextFilters, vd, alwaysFireEvent);
+                    });
+                    return Promise.resolve(doExchange);
+                })
+                .catch_(err -> {
+                    failureHandled(err.toString());
+                    return null;
+                });
+    }
 
-        final RangeSet includedRows;
-        if (result.getSnapshotEnd() < result.getSnapshotStart()) {
-            // this is TSQ for "no items"
-            includedRows = RangeSet.empty();
-        } else {
-            includedRows = RangeSet.ofRange(result.getSnapshotStart(), result.getSnapshotEnd());
-        }
+    /**
+     * Instead of a micro-task between chained promises, insert a regular task so that control is returned to the
+     * browser long enough to prevent the UI hanging.
+     */
+    private <T> IThenable.ThenOnFulfilledCallbackFn<T, T> defer() {
+        return val -> new Promise<>((resolve, reject) -> {
+            DomGlobal.setTimeout(ignoreArgs -> resolve.onInvoke(val), 0);
+        });
+    }
 
-        TreeViewportData vd = new TreeViewportData(
-                includedRows,
-                result.getSnapshotData(),
-                result.getKeyColumn(),
-                result.getParentKeyColumn(),
-                result.getChildPresence(),
-                result.getSnapshotStart(),
-                columns,
-                baseTable.getRowFormatColumn() == null ? NO_ROW_FORMAT_COLUMN
-                        : baseTable.getRowFormatColumn().getIndex(),
-                result.getConstituentColumnNames(),
-                result.getConstituentColumnData());
+    private void handleUpdate(List<Sort> nextSort, List<FilterCondition> nextFilters,
+            TreeViewportData viewportData, boolean alwaysFireEvent) {
+        JsLog.debug("tree table response arrived", viewportData);
+        if (closed) {
+            // ignore
+            return;
+        }
 
         // if requested to fire the event, or if the data has changed in some way, fire the event
-        final boolean fireEvent = alwaysFireEvent || !vd.containsSameDataAs(currentViewportData);
-
-        this.currentViewportData = vd;
-
-        Set<Key> presentKeys = new HashSet<>();
-        for (int i = 0; i < result.getTableDetails().length; i++) {
-            TableDetails detail = result.getTableDetails()[i];
-            TreeNodeState treeNodeState = expandedMap.get(detail.getKey());
+        final boolean fireEvent = true;// alwaysFireEvent || !viewportData.containsSameDataAs(currentViewportData);
```

Comment:
Revert this line
```suggestion
        final boolean fireEvent = alwaysFireEvent || !viewportData.containsSameDataAs(currentViewportData);
```

Reply:
Okay cool can you remove the comment after then and capture in a TODO then?

Reply:
Done, #3310